storage/database/sharded_database.go

```diff
@@ -151,102 +154,280 @@ func (pdb *shardedDB) Close() {
+func (db *shardedDB) NewIteratorWithStartUnsorted(start []byte) Iterator {
+	return db.newIteratorUnsorted(func(db Database) Iterator { return db.NewIteratorWithStart(start) })
+}
+
+// NewIteratorWithPrefixUnsorted creates a iterator over a subset of database content
+// with a particular key prefix.
+func (db *shardedDB) NewIteratorWithPrefixUnsorted(prefix []byte) Iterator {
+	return db.newIteratorUnsorted(func(db Database) Iterator { return db.NewIteratorWithPrefix(prefix) })
+}
+
+func (db *shardedDB) newIteratorUnsorted(newIterator func(Database) Iterator) Iterator {
+	it := &shardedDBIteratorUnsorted{
+		shardedDBIterator{
+			db.NewChanIterator(context.Background(), newIterator),
+			make(chan entry, shardedDBCombineChanSize),
+			nil, nil}}
+
+	go it.newCombineWorker()
+
+	return it
+}
+
+// runCombineWorker fetches any key/value from resultChs and put the data in resultCh
+func (it shardedDBIteratorUnsorted) newCombineWorker() {
+Iter:
+	for len(it.resultChs) != 0 {
+	chanIter:
+		for i, ch := range it.resultChs {
+			select {
+			case <-it.ctx.Done():
+				logger.Trace("[shardedDBIteratorUnsorted] combine worker ends due to ctx")
+				break Iter
+			case e, ok := <-ch:
+				if !ok { // channel is closed
+					it.resultChs = append(it.resultChs[:i], it.resultChs[i+1:]...)
+					break chanIter
+				} else {
+					it.resultCh <- e
+				}
+			}
+		}
+	}
+	logger.Trace("[shardedDBIteratorUnsorted] combine worker finishes")
+	close(it.resultCh)
+}
+
+// shardedDBChanIterator creates iterators for each shard DB.
+// Channels subscribing each iterators can be gained.
+// Each iterators fetch values in binary-alphabetical order.
+// This is useful when you want to operate on each items in parallel.
+type shardedDBChanIterator struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	iterators []Iterator
+	resultChs []chan entry
+}
+
+// NewChanIterator creates iterators for each shard DB.
+// This is useful when you want to operate on each items in parallel.
+// If you want to get ordered items in serial, checkout shardedDB.NewIterator()
+// If you want to get unordered items in serial, checkout shardedDB.NewChanIteratorUnsorted().
+func (db *shardedDB) NewChanIterator(ctx context.Context, newIterator func(Database) Iterator) shardedDBChanIterator {
+	if ctx == nil {
+		ctx = context.Background()
+	}
+
+	it := shardedDBChanIterator{ctx, nil,
+		make([]Iterator, len(db.shards)),
+		make([]chan entry, len(db.shards))}
+	it.ctx, it.cancel = context.WithCancel(ctx)
+
+	for i := 0; i < len(db.shards); i++ {
+		it.iterators[i] = newIterator(db.shards[i])
+		it.resultChs[i] = make(chan entry, shardedDBSubChannelSize)
+		go it.newChanWorker(it.iterators[i], it.resultChs[i], it.ctx)
+	}
+
+	return it
+}
+
+func (*shardedDBChanIterator) newChanWorker(it Iterator, resultCh chan entry, ctx context.Context) {
+iter:
+	for it.Next() {
+		select {
+		case <-ctx.Done():
+			break iter
+		default:
+		}
```

Comment:
if child DB iteratorcan get `cancle()`, this routine is not necessary?

Reply:
I tried to use `Release()` and remove `context` and `cancel`. However, when I do, I need a lock, instead. A LevelDB iterator is not for concurrent access. A channel iterator keeps calling `Next()` and calling `Release()` on other go routine makes an `nil point error`. 
Using a lock is more complicated than a context. Moreover, context can also handle process exit and user's cancellation signal. @ehnuje